paywall/src/components/lock/NoKeyLock.js

```diff
@@ -5,64 +5,74 @@ import UnlockPropTypes from '../../propTypes'
 import { LockWrapper, LockHeader, LockBody, LockFooter } from './LockStyles'
 import BalanceProvider from '../helpers/BalanceProvider'
 import Duration from '../helpers/Duration'
+import { UNLIMITED_KEYS_COUNT } from '../../constants'
 
 export const NoKeyLock = ({
+  account,
   lock,
   disabled,
   purchaseKey,
   lockKey,
-  soldOut,
-  tooExpensive,
-}) => (
-  <Wrapper
-    lock={lock}
-    disabled={disabled}
-    onClick={() => {
-      !disabled && purchaseKey(lockKey)
-    }}
-  >
-    <LockHeader>{lock.name}</LockHeader>
-    <BalanceProvider
-      amount={lock.keyPrice}
-      render={(ethPrice, fiatPrice) => (
-        <div>
-          <Body disabled={disabled}>
-            <EthPrice>{ethPrice} Eth</EthPrice>
-            <div>
-              <FiatPrice>${fiatPrice}</FiatPrice>
-              <Separator> | </Separator>
-              <ExpirationDuration>
-                <Duration seconds={lock.expirationDuration} round />
-              </ExpirationDuration>
-            </div>
-            <Footer>
-              {soldOut
-                ? 'Sold Out'
-                : tooExpensive
-                ? 'Insufficient funds'
-                : 'Purchase'}
-            </Footer>
-          </Body>
-        </div>
-      )}
-    />
-  </Wrapper>
-)
+}) => {
+  const soldOut =
+    lock.outstandingKeys >= lock.maxNumberOfKeys &&
+    lock.maxNumberOfKeys !== UNLIMITED_KEYS_COUNT
+
+  // When the lock is not disabled for other reasons (pending key on
+  // other lock...), we need to ensure that the lock is disabled
+  // when the lock is sold out or too expensive for the current account
+  const tooExpensive =
+    account && parseFloat(account.balance) <= parseFloat(lock.keyPrice)
+
+  return (
+    <Wrapper
+      lock={lock}
+      disabled={disabled || tooExpensive || soldOut}
```

Comment:
another opportunity - perhaps we define this above so all our computation is above the render? This is not super important, but since you're refactoring anyways :)

Reply:
It's actually more complicated than that.
This component should have the logic of `tooExpensive` and `soldOut`... but a lock can be disabled for other reasons (especially if there is another lock with a pending key for example)

Reply:
sorry! I was unclear. I meant something like:

```js
const disableView = disabled || tooExpensive || soldOut

...

    <Wrapper
      lock={lock}
      disabled={disableView}
```

Reply:
oh, ok? Is that really better?

Reply:
it isn't better per se, just congregates all the logic in the top, and the display in the bottom, for readability